mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp

```diff
@@ -840,6 +840,11 @@ LogicalResult RegionPatternRewriteDriver::simplify(bool *changed) && {
     // regions to enable more aggressive CSE'ing).
     OperationFolder folder(ctx, this);
     auto insertKnownConstant = [&](Operation *op) {
+      // This hoisting is to enable more folding, so skip checking if known
+      // constant, updating dense map etc if not doing folding.
+      if (!config.fold)
```

Comment:
`insertKnownConstant` is not really doing "folding": it does not call the folder. What it does is CSE'ing constants. So I would go as far as adding a second config flag: `cseConstants` (see #89552). Alternatively, update the `GreedyRewriteConfig::fold` documentation and mention that setting it to "false" also disables the CSE'ing of constants.

Reply:
I am quite concern with the accumulation of flags: this seems to me like working around problems instead of addressing them, and adding complexity that I don't see manageable on the long term.

Reply:
What problem do you think is being worked around?

For me I see this as I'm a user that wants a worklist driven driver, I have a curated pipeline where I know there isn't any folding here (e.g., the ops being created have no folding, its part of pipeline where folding is ~0% likely, so the ROI on checking if it can be folded is none), but they have no way of communicating that intent. The problem is expert users not being able to express intent without crafting own/local forking.

Now, constant CSE'ing as Matthias correctly points out is different. It is documented here as enabling more folding, so disabled too (this is cheaper here too as all constant hoisting would have happened earlier and there may be no additional chance here to produce additional constants that would need CSE'ing). But it is different so I could see controlling separately is good. And don't mind doing that. I think this is a framework for high performance, so reasonably defaults that can be tuned seems reasonable to me. I considered the maintenance cost but specialized drivers that share most of the same code seemed higher maintenance cost to me.